iroh/src/downloader.rs

```diff
@@ -0,0 +1,979 @@
+    current_requests: HashMap<DownloadKind, ActiveRequestInfo>,
+    /// Downloads underway.
+    in_progress_downloads: FuturesUnordered<DownloadFut>,
+    /// Requests scheduled to be downloaded at a later time.
+    scheduled_requests: HashMap<DownloadKind, PendingRequestInfo>,
+    /// Queue of scheduled requests.
+    scheduled_request_queue: delay_queue::DelayQueue<DownloadKind>,
+}
+
+impl<G: Getter<Connection = D::Connection>, D: Dialer> Service<G, D> {
+    fn new(
+        getter: G,
+        dialer: D,
+        concurrency_limits: ConcurrencyLimits,
+        msg_rx: mpsc::Receiver<Message>,
+    ) -> Self {
+        Service {
+            getter,
+            providers: ProviderMap::default(),
+            dialer,
+            concurrency_limits,
+            msg_rx,
+            peers: HashMap::default(),
+            goodbye_peer_queue: delay_queue::DelayQueue::default(),
+            current_requests: HashMap::default(),
+            in_progress_downloads: FuturesUnordered::default(),
+            scheduled_requests: HashMap::default(),
+            scheduled_request_queue: delay_queue::DelayQueue::default(),
+        }
+    }
+
+    /// Main loop for the service.
+    async fn run(mut self) {
+        loop {
+            // check if we have capacity to dequeue another scheduled request
+            let at_capacity = self
+                .concurrency_limits
+                .at_requests_capacity(self.in_progress_downloads.len());
+
+            tokio::select! {
+                Some((peer, conn_result)) = self.dialer.next() => {
+                    trace!("tick: connection ready");
+                    self.on_connection_ready(peer, conn_result);
+                }
+                maybe_msg = self.msg_rx.recv() => {
+                    trace!(msg=?maybe_msg, "tick: message received");
+                    match maybe_msg {
+                        Some(msg) => self.handle_message(msg),
+                        None => return self.shutdown().await,
+                    }
+                }
+                Some((kind, result)) = self.in_progress_downloads.next() => {
+                    trace!("tick: download completed");
+                    self.on_download_completed(kind, result);
+                }
+                Some(expired) = self.scheduled_request_queue.next(), if !at_capacity => {
+                    trace!("tick: scheduled request ready");
+                    let kind = expired.into_inner();
+                    let request_info = self.scheduled_requests.remove(&kind).expect("is registered");
+                    self.on_scheduled_request_ready(kind, request_info);
+                }
+                Some(expired) = self.goodbye_peer_queue.next() => {
+                    let peer = expired.into_inner();
+                    self.peers.remove(&peer);
+                    trace!(%peer, "tick: goodbye peer");
+                }
+            }
+            #[cfg(test)]
```

Comment:
maybe `cfg(debug_assertions)`?

Reply:
sounds good, tho that would require me to mos the function somewhere else since it's currently under test

But yeah agree

Reply:
done in 7e704d73